# **Northwind Traders Sales Analysis Project**

![Sales Analysis](sales.jpg)

Welcome to the Northwind Traders Sales Analysis Project! This project aims to provide comprehensive insights into the sales performance, customer behavior, and product trends of Northwind Traders. Through this analysis, we will achieve the following:

- **Sales Performance Analysis:** Understand the overall sales trends and identify the top-performing products and customers.
- **Customer Behavior Analysis:** Explore customer purchasing patterns and determine the most profitable customer segments.
- **Product Trends Analysis:** Examine product popularity, category performance, and supplier contributions.

By the end of this project, we will provide actionable insights to help Northwind Traders enhance their business strategies and improve sales performance.

## **Project Outline**

1. **Objective/Goal Definition**
2. **Data Exploration and Understanding**
3. **Formulating Questions**
4. **Data Analysis and Querying**
5. **Data Visualization**
6. **Insights and Conclusions**

Let's dive into the data and start uncovering valuable insights!


In [1]:
%load_ext sql

%sql postgresql://postgres:pranshul@localhost:5432/northwind


In [2]:
import psycopg2
import pandas as pd 
import warnings
warnings.filterwarnings('ignore')

host = "localhost"
database = "northwind"
user = "postgres"
password = "pranshul"

# Connect to the PostgreSQL database
conn = psycopg2.connect(
    host=host,
    database=database,
    user=user,
    password=password
)

In [3]:
import matplotlib.pyplot as plt
import seaborn as sns 
import plotly.express as px

In [4]:
%%sql 
-- Create view
CREATE VIEW table_columns AS 
SELECT
    table_name,
    STRING_AGG(column_name,', ') AS columns
FROM information_schema.columns
WHERE table_schema = 'public'
GROUP BY table_name

 * postgresql://postgres:***@localhost:5432/northwind
(psycopg2.errors.DuplicateTable) relation "table_columns" already exists

[SQL: -- Create view
CREATE VIEW table_columns AS 
SELECT
    table_name,
    STRING_AGG(column_name,', ') AS columns
FROM information_schema.columns
WHERE table_schema = 'public'
GROUP BY table_name]
(Background on this error at: https://sqlalche.me/e/20/f405)


In [5]:
%%sql
-- USING view
SELECT * FROM table_columns

 * postgresql://postgres:***@localhost:5432/northwind
15 rows affected.


[('suppliers', 'supplier_id, company_name, contact_name, contact_title, address, city, region, postal_code, country, phone, fax, homepage'),
 ('customer_demographics', 'customer_type_id, customer_desc'),
 ('territories', 'region_id, territory_id, territory_description'),
 ('shippers', 'shipper_id, company_name, phone'),
 ('orders', 'order_id, employee_id, order_date, required_date, shipped_date, ship_via, freight, customer_id, ship_name, ship_address, ship_city, ship_region, ship_postal_code, ship_country'),
 ('customer_customer_demo', 'customer_id, customer_type_id'),
 ('order_details', 'discount, order_id, product_id, unit_price, quantity'),
 ('employees', 'employee_id, birth_date, hire_date, photo, reports_to, home_phone, extension, notes, photo_path, last_name, first_name, title, title_of_courtesy, address, city, region, postal_code, country'),
 ('table_columns', 'table_name, columns'),
 ('categories', 'category_id, picture, description, category_name'),
 ('employee_territories', 'employee_id, territory_id'),
 ('customers', 'customer_id, company_name, contact_name, contact_title, address, city, region, postal_code, country, phone, fax'),
 ('products', 'product_id, supplier_id, category_id, unit_price, units_in_stock, units_on_order, reorder_level, discontinued, product_name, quantity_per_unit'),
 ('region', 'region_id, region_description'),
 ('us_states', 'state_id, state_name, state_abbr, state_region')]

## Sales Performance

### What are the total sales by year?

In [78]:
query = """
SELECT
    EXTRACT(YEAR FROM o.order_date)::INT AS Year,
    SUM(od.unit_price * od.quantity * (1 - od.discount)) AS TotalSales
FROM
    orders o
JOIN
    Order_details od ON o.order_id = od.order_id
GROUP BY
    EXTRACT(YEAR FROM o.order_date)
ORDER BY
    Year
"""

df = pd.read_sql_query(query, conn)
df

year     totalsales
0  1996  208083.970983
1  1997  617085.202393
2  1998  440623.865278

In [79]:
# Plotting
fig = px.bar(df, x='year', y='totalsales', title='Total Sales by Year',
             labels={'year': 'Year', 'totalsales': 'Total Sales'},
             template='plotly_dark')
fig.show()


**Findings:**
The analysis of total sales by year reveals the following:
- **1996:** Total sales amounted to approximately $208,084.

- **1997:** Total sales increased significantly to around $617,085, indicating a strong growth year.

- **1998:** Sales slightly decreased to approximately $440,624 but remained substantial.

**Visualization:**
A bar graph of the total sales by year clearly shows a peak in 1997, suggesting this year was exceptionally strong in terms of sales performance. The subsequent decline in 1998 might warrant further investigation to understand potential causes.

**Insight:**
The spike in 1997 could be attributed to various factors such as successful marketing campaigns, product launches, or market conditions. The drop in 1998 could indicate the need for a deeper analysis into factors affecting sales, such as market saturation or competitive dynamics.


### Which products generate the most revenue?

In [85]:
query = """
SELECT
    p.product_name,
    ROUND(SUM(od.unit_price * od.quantity * (1 - od.discount))::NUMERIC,2) AS TotalRevenue
FROM
    order_details AS od
JOIN
    products AS p ON od.product_id = p.product_id
GROUP BY
    p.product_name
ORDER BY
    TotalRevenue DESC
"""
df = pd.read_sql_query(query,conn)
df.head(10)


product_name  totalrevenue
0            Côte de Blaye     141396.74
1  Thüringer Rostbratwurst      80368.67
2     Raclette Courdavault      71155.70
3           Tarte au sucre      47234.97
4        Camembert Pierrot      46825.48
5   Gnocchi di nonna Alice      42593.06
6    Manjimup Dried Apples      41819.65
7             Alice Mutton      32698.38
8         Carnarvon Tigers      29171.87
9        Rössle Sauerkraut      25696.64

In [81]:
#Plotting
fig = px.bar(df.head(10), x='product_name', y='totalrevenue', title='Top 10 Products generates high Revenue',
            labels={'product_name':'Product Name','totalrevenue':"Total Revenue"},
            template='plotly_dark',
            color='totalrevenue')

# Rotate x-axis labels for better readability
fig.update_layout(xaxis_title='Product Name',yaxis_title='Total Revenue',
                 xaxis_tickangle=-45)
fig.show()

In [87]:
#Plotting
fig = px.bar(df.tail(10), x='product_name', y='totalrevenue', title='bottom 10 Products generates Lowest Revenue',
            labels={'product_name':'Product Name','totalrevenue':"Total Revenue"},
            template='plotly_dark',
            color='totalrevenue')

# Rotate x-axis labels for better readability
fig.update_layout(xaxis_title='Product Name',yaxis_title='Total Revenue',
                 xaxis_tickangle=-45)
fig.show()

### Which Products Generate the Most Revenue?

**Objective:** Identify which products generate the highest and lowest revenue to understand their financial impact.

**Findings:**

- **Top Products by Revenue:**
  - **Côte de Blaye:** Generated the highest revenue of approximately $141,397.
  
  - **Thüringer Rostbratwurst:** Followed with a revenue of around $80,369.
  
  - **Raclette Courdavault:** Contributed $71,156 in revenue.
  
  - **Tarte au sucre:** Brought in about $47,235.
  
  - **Camembert Pierrot:** Earned approximately $46,825.
  

  The top 10 products significantly impact overall revenue, with the leading products driving substantial earnings.

- **Bottom Products by Revenue:**
  - **Chocolade:** Generated the lowest revenue at around $1,369.
  
  - **Geitost:** Followed with a revenue of approximately $1,648.
  
  - **Genen Shouyu:** Brought in about $1,785.
  
  - **Laughing Lumberjack Lager:** Earned around $2,397.
  
  - **Longlife Tofu:** Contributed $2,433.

  The bottom 10 products contribute less to the total revenue, highlighting their limited financial impact.

**Visualization:**
A bar graph for the top 10 products visually represents the highest revenue-generating items, while the bottom 10 products can be further analyzed to explore their lower revenue figures.

**Insight:**
The top revenue-generating products are crucial to the business's financial health and should be prioritized. Conversely, the bottom revenue-generating products may need strategic review to understand their limited performance and potential for improvement.


### Who are the top customers by sales amount?

In [10]:
query = '''
SELECT
    c.customer_id,
    c.company_name,
    ROUND(SUM(od.unit_price * quantity)::NUMERIC,2) AS TotalSales
FROM 
    orders AS o 
JOIN order_details AS od USING(order_id)
JOIN customers AS c 
ON o.customer_id = c.customer_id
GROUP BY c.customer_id, c.company_name
ORDER BY TotalSales DESC'''

df = pd.read_sql_query(query,conn)
df.head(10)

customer_id                  company_name  totalsales
0       QUICK                    QUICK-Stop   117483.39
1       SAVEA            Save-a-lot Markets   115673.39
2       ERNSH                  Ernst Handel   113236.68
3       HUNGO  Hungry Owl All-Night Grocers    57317.39
4       RATTC    Rattlesnake Canyon Grocery    52245.90
5       HANAR                 Hanari Carnes    34101.15
6       FOLKO                Folk och fä HB    32555.55
7       MEREP                Mère Paillarde    32203.90
8       KOENE               Königlich Essen    31745.75
9       QUEEN                 Queen Cozinha    30226.10

In [11]:
#Plotting
fig = px.bar(df.head(10), x='customer_id', y='totalsales', title='Top 10 Customers generates high sales',
            labels={"customer_id":"Customer ID",'totalsales':'Total Sales'},
            template='plotly_dark',
            color='totalsales')

# Rotate x-axis labels for better readability
fig.update_layout(xaxis_title='Customer Id', yaxis_title='Total Sales',
                 xaxis_tickangle=-45)
fig.show()

### Who Are the Top Customers by Sales Amount?

**Objective:** Identify the top customers based on their total sales amount to understand who contributes the most to revenue.

**Findings:**

- **Top Customers by Sales Amount:**

  - **QUICK-Stop:** The top customer with a total sales amount of approximately $117,483.
  
  - **Save-a-lot Markets:** Contributed around $115,673 in sales.
  
  - **Ernst Handel:** Generated $113,237 in total sales.
  
  - **Hungry Owl All-Night Grocers:** Brought in about $57,317.
  
  - **Rattlesnake Canyon Grocery:** Accounted for $52,246 in sales.

  The top 10 customers significantly impact overall sales, indicating their importance to the business’s revenue stream.

**Visualization:**
A bar graph of the top 10 customers visually highlights those who contribute the most to sales, providing a clear comparison of their spending levels.

**Insight:**
These top customers are key to the company’s revenue and should be prioritized for relationship management and targeted marketing efforts. Understanding their purchasing behavior can help in maintaining and growing these valuable accounts.


### What are the sales trends by region?

In [12]:
query = '''
SELECT
    o.ship_country AS Region,
    SUM(od.unit_price * od.quantity) AS TotalSales
FROM
    orders AS o
JOIN
    order_details AS od ON o.order_id = od.order_id
GROUP BY
    o.ship_country
ORDER BY
    TotalSales DESC

'''
df = pd.read_sql_query(query,conn)
df

region     totalsales
0           USA  263566.980017
1       Germany  244640.629969
2       Austria  139496.629867
3        Brazil  114968.480040
4        France   85498.760092
5     Venezuela   60814.889832
6            UK   60616.509948
7        Sweden   59523.699950
8       Ireland   57317.390162
9        Canada   55334.100187
10      Belgium   35134.980321
11      Denmark   34782.250007
12  Switzerland   32919.499998
13       Mexico   24073.449984
14      Finland   19778.450103
15        Spain   19431.889992
16        Italy   16705.149927
17     Portugal   12468.650044
18    Argentina    8119.099991
19       Norway    5735.150015
20       Poland    3531.949996

In [13]:
#Plotting
fig = px.bar(df, x='region', y='totalsales', title='Sales Trend by Region',
            labels={'region':"Region",'totalsales':"Total Sales"},
            template='plotly_dark',
            color='totalsales')
fig.show()

### What Are the Sales Trends by Region?

**Objective:** Analyze the total sales amounts by region to identify which regions contribute most to sales and understand regional performance trends.

**Findings:**

- **Top Regions by Sales:**

  - **USA:** Leads with the highest total sales of approximately $263,567.
  
  - **Germany:** Follows with significant sales totaling around $244,641.
  
  - **Austria:** Contributed $139,497 in sales.
  
  - **Brazil:** Generated $114,968 in total sales.
  
  - **France:** Brought in about $85,499.

  The top regions show high sales performance, indicating strong market presence and customer demand in these areas.

- **Lower Sales Regions:**

  - **Poland:** Recorded the lowest total sales at around $3,532.
  
  - **Norway:** Generated $5,735 in sales.
  
  - **Argentina:** Contributed $8,119.
  
  - **Portugal:** Brought in about $12,469.
  
  - **Italy:** Had total sales of $16,705.

  The lower sales regions contribute less to the overall revenue, highlighting potential areas for growth or further investigation.

**Visualization:**
The bar graph depicting sales by region provides a clear comparison of sales performance across different regions, showcasing the disparity in sales volumes.

**Insight:**
The analysis indicates that the USA and Germany are key markets with substantial revenue contributions. Regions with lower sales may benefit from targeted strategies to enhance market presence and drive growth.


In [15]:

%%sql
SELECT * FROM table_columns

 * postgresql://postgres:***@localhost:5432/northwind
15 rows affected.


[('suppliers', 'supplier_id, company_name, contact_name, contact_title, address, city, region, postal_code, country, phone, fax, homepage'),
 ('customer_demographics', 'customer_type_id, customer_desc'),
 ('territories', 'region_id, territory_id, territory_description'),
 ('shippers', 'shipper_id, company_name, phone'),
 ('orders', 'order_id, employee_id, order_date, required_date, shipped_date, ship_via, freight, customer_id, ship_name, ship_address, ship_city, ship_region, ship_postal_code, ship_country'),
 ('customer_customer_demo', 'customer_id, customer_type_id'),
 ('order_details', 'discount, order_id, product_id, unit_price, quantity'),
 ('employees', 'employee_id, birth_date, hire_date, photo, reports_to, home_phone, extension, notes, photo_path, last_name, first_name, title, title_of_courtesy, address, city, region, postal_code, country'),
 ('table_columns', 'table_name, columns'),
 ('categories', 'category_id, picture, description, category_name'),
 ('employee_territories', 'employee_id, territory_id'),
 ('customers', 'customer_id, company_name, contact_name, contact_title, address, city, region, postal_code, country, phone, fax'),
 ('products', 'product_id, supplier_id, category_id, unit_price, units_in_stock, units_on_order, reorder_level, discontinued, product_name, quantity_per_unit'),
 ('region', 'region_id, region_description'),
 ('us_states', 'state_id, state_name, state_abbr, state_region')]

## Customer Behavior

### How frequently do customers place orders?

In [25]:

query = """SELECT 
    customer_id,
    COUNT(order_id) AS order_count
FROM customers AS c
JOIN orders AS s 
USING(customer_id)
JOIN order_details
USING(order_id)
GROUP BY customer_id
ORDER BY order_count DESC
"""
df = pd.read_sql_query(query,conn)
df


customer_id  order_count
0        SAVEA          116
1        ERNSH          102
2        QUICK           86
3        RATTC           71
4        HUNGO           55
..         ...          ...
84       FRANR            6
85       SPECD            6
86       GROSR            4
87       CENTC            2
88       LAZYK            2

[89 rows x 2 columns]

In [26]:
# plotting 
fig = px.histogram(df,
                  x='order_count',
                  nbins=10,
                  title='Order Frequency Distribution',
                  labels={"order_cunt":"Number of Orders"},
                  template='plotly_dark')
fig.show()

### How Frequently Do Customers Place Orders?

**Objective:** Determine how often customers place orders to understand their purchasing behavior and identify high-frequency customers.

**Findings:**

- **Top Customers by Order Frequency:**
    
  - **Save-a-lot Markets (SAVEA):** Placed the most orders with a total of 116.
    
  - **Ernst Handel (ERNSH):** Followed with 102 orders.
    
  - **QUICK-Stop (QUICK):** Made 86 orders.
    
  - **Rattlesnake Canyon Grocery (RATTC):** Had 71 orders.
    
  - **Hungry Owl All-Night Grocers (HUNGO):** Placed 55 orders.

  These top customers exhibit frequent ordering behavior, indicating strong and consistent purchasing patterns.

- **Overall Trends:**
    
  - The histogram of order frequency shows the distribution of how often customers place orders, highlighting the frequency of purchases across the customer base.

**Visualization:**
    
The histogram provides a visual representation of order frequency, illustrating how many orders customers place and helping identify high-frequency buyers.

**Insight:**
    
High-frequency customers are valuable and indicate strong engagement with the company. Understanding their ordering behavior can help in tailoring marketing efforts and maintaining customer loyalty.


### What is the average order value per customer?

In [90]:

query = """SELECT customer_id, AVG(unit_price * quantity * (1-discount))
FROM customers AS c 
JOIN orders AS o 
USING(customer_id)
JOIN order_details AS od 
USING(order_id)
GROUP BY customer_id
ORDER BY avg DESC
LIMIT 10
"""

df = pd.read_sql_query(query,conn)
df

customer_id          avg
0       QUICK  1282.294245
1       SIMOB  1121.139841
2       ERNSH  1028.186060
3       HANAR  1026.292811
4       PICCO  1005.602609
5       HUNGO   908.725547
6       MEREP   902.255942
7       SAVEA   899.671979
8       GREAL   841.247726
9       KOENE   792.522663

In [91]:
# visualizing top 10 customers with highest price average
fig = px.bar(df,
            x='customer_id',
            y='avg',
            title='Average Order Value of Top 10 Customers',
            labels={"customer_id":"Customer ID","avg":"Average Order Value"},
            color='avg',
            color_continuous_scale='Viridis',
            template='plotly_dark')

fig.show()



**Findings:**

- **Top Customers by Average Order Value (After Discount):**

  - **QUICK-Stop (QUICK):** Has the highest average order value at approximately $1,282.29.
  
  - **Simob (SIMOB):** Follows with an average order value of around $1,121.14.
  
  - **Ernst Handel (ERNSH):** Contributed an average of $1,028.19 per order.
  
  - **Hanari Carnes (HANAR):** Averaged $1,026.29 per order.
  
  - **Picco (PICCO):** Recorded an average order value of $1,005.60.

  These top customers show higher spending per order even after discounts, indicating their significant purchasing power.

- **Overall Trends:**
  - The bar graph visualizes the average order value per customer, showing the impact of discounts on spending and highlighting high-value customers.

**Visualization:**
The bar graph illustrates the average order value for the top 10 customers, providing a clear view of their spending patterns.

**Insight:**
Customers with higher average order values are important for revenue generation. Tailoring marketing strategies to these high-value customers can further boost their engagement and overall sales.


In [57]:

query = '''SELECT
    c.customer_id,
    c.company_name,
    SUM(od.unit_price * od.quantity * (1 - od.discount)) AS TotalSales,
    RANK() OVER(ORDER BY SUM(od.unit_price * od.quantity * (1 - od.discount)) DESC)
FROM
    orders o
JOIN
    order_details od ON o.order_id = od.order_id
JOIN
    customers c ON o.customer_id = c.customer_id
GROUP BY
    c.customer_id, c.company_name
ORDER BY
    TotalSales DESC
LIMIT 10'''
df = pd.read_sql_query(query,conn)
df

customer_id                  company_name     totalsales  rank
0       QUICK                    QUICK-Stop  110277.305030     1
1       ERNSH                  Ernst Handel  104874.978144     2
2       SAVEA            Save-a-lot Markets  104361.949540     3
3       RATTC    Rattlesnake Canyon Grocery   51097.800828     4
4       HUNGO  Hungry Owl All-Night Grocers   49979.905081     5
5       HANAR                 Hanari Carnes   32841.369948     6
6       KOENE               Königlich Essen   30908.383873     7
7       FOLKO                Folk och fä HB   29567.562490     8
8       MEREP                Mère Paillarde   28872.190156     9
9       WHITC          White Clover Markets   27363.604900    10

In [61]:
## Top 10 customers with highest sales
fig = px.bar(df,
            x='customer_id',
            y='totalsales',
            title='Top 10 Customers contributing highest sales',
            labels={"customer_id":"Customer ID",'totalsales':"Total Sales"},
            template='plotly_dark')
fig.show()

### Product Trends

### What are the most and least popular products

#### In terms of total_sales

In [73]:
%%sql
-- Most pop products
SELECT
    p.product_id,
    p.product_name,
    SUM(od.unit_price * od.quantity * (1 - od.discount)) AS TotalSales
FROM
    order_details od
JOIN
    products p ON od.product_id = p.product_id
GROUP BY
    p.product_id, p.product_name
ORDER BY
    TotalSales DESC
LIMIT 10;


 * postgresql://postgres:***@localhost:5432/northwind
10 rows affected.


[(38, 'Côte de Blaye', 141396.7356273254),
 (29, 'Thüringer Rostbratwurst', 80368.6724385033),
 (59, 'Raclette Courdavault', 71155.69990943),
 (62, 'Tarte au sucre', 47234.969978504174),
 (60, 'Camembert Pierrot', 46825.48029542655),
 (56, 'Gnocchi di nonna Alice', 42593.0598222503),
 (51, 'Manjimup Dried Apples', 41819.65024582073),
 (17, 'Alice Mutton', 32698.380216373203),
 (18, 'Carnarvon Tigers', 29171.874963399023),
 (28, 'Rössle Sauerkraut', 25696.63978933155)]

In [92]:
%%sql
-- least popular products
SELECT
    p.product_id,
    p.product_name,
    SUM(od.unit_price * od.quantity * (1 - od.discount)) AS TotalSales
FROM
    order_details od
JOIN
    products p ON od.product_id = p.product_id
GROUP BY
    p.product_id, p.product_name
ORDER BY
    TotalSales 
LIMIT 10;


 * postgresql://postgres:***@localhost:5432/northwind
10 rows affected.


[(48, 'Chocolade', 1368.7124797306956),
 (33, 'Geitost', 1648.1249991375953),
 (15, 'Genen Shouyu', 1784.824990035966),
 (67, 'Laughing Lumberjack Lager', 2396.799994945526),
 (74, 'Longlife Tofu', 2432.4999990463257),
 (37, 'Gravad lax', 2688.3999648988247),
 (3, 'Aniseed Syrup', 3043.999999463558),
 (52, 'Filo Mix', 3232.949990465492),
 (66, 'Louisiana Hot Spiced Okra', 3383.000052154064),
 (50, 'Valkoinen suklaa', 3437.68749892246)]

#### In terms of quantity

In [74]:
%%sql
-- MOST
SELECT
    p.product_id,
    p.product_name,
    SUM(od.quantity) AS TotalQuantitySold
FROM
    order_details od
JOIN
    products p ON od.product_id = p.product_id
GROUP BY
    p.product_id, p.product_name
ORDER BY
    TotalQuantitySold DESC
LIMIT 10;


 * postgresql://postgres:***@localhost:5432/northwind
10 rows affected.


[(60, 'Camembert Pierrot', 1577),
 (59, 'Raclette Courdavault', 1496),
 (31, 'Gorgonzola Telino', 1397),
 (56, 'Gnocchi di nonna Alice', 1263),
 (16, 'Pavlova', 1158),
 (75, 'Rhönbräu Klosterbier', 1155),
 (24, 'Guaraná Fantástica', 1125),
 (40, 'Boston Crab Meat', 1103),
 (62, 'Tarte au sucre', 1083),
 (2, 'Chang', 1057)]

In [93]:
%%sql
-- LEAST
SELECT
    p.product_id,
    p.product_name,
    SUM(od.quantity) AS TotalQuantitySold
FROM
    order_details od
JOIN
    products p ON od.product_id = p.product_id
GROUP BY
    p.product_id, p.product_name
ORDER BY
    TotalQuantitySold 
LIMIT 10;


 * postgresql://postgres:***@localhost:5432/northwind
10 rows affected.


[(9, 'Mishi Kobe Niku', 95),
 (15, 'Genen Shouyu', 122),
 (37, 'Gravad lax', 125),
 (48, 'Chocolade', 138),
 (67, 'Laughing Lumberjack Lager', 184),
 (50, 'Valkoinen suklaa', 235),
 (66, 'Louisiana Hot Spiced Okra', 239),
 (73, 'Röd Kaviar', 293),
 (74, 'Longlife Tofu', 297),
 (32, 'Mascarpone Fabioli', 297)]

### What Are the Most and Least Popular Products in Terms of Total Sales and Quantity?

**Objective:** Identify the most and least popular products based on total sales and quantity sold to understand their market performance.

**Findings:**

- **Most Popular Products by Total Sales:**
  - **Côte de Blaye:** Highest total sales of approximately $141,397.
    
  - **Thüringer Rostbratwurst:** Significant sales of about $80,369.
    
  - **Raclette Courdavault:** Generated approximately $71,156 in sales.
    
  - **Tarte au sucre:** Total sales of around $47,235.
    
  - **Camembert Pierrot:** Total sales of about $46,825.

- **Least Popular Products by Total Sales:**
  - **Chocolade:** Lowest sales at approximately $1,369.
    
  - **Geitost:** Sales of around $1,648.
    
  - **Genen Shouyu:** Total sales of about $1,785.
    
  - **Laughing Lumberjack Lager:** Generated about $2,397.
    
  - **Longlife Tofu:** Total sales of approximately $2,432.

- **Most Popular Products by Quantity Sold:**
  - **Camembert Pierrot:** Highest quantity sold at 1,577 units.
    
  - **Raclette Courdavault:** Sold 1,496 units.
    
  - **Gorgonzola Telino:** Total quantity sold of 1,397 units.
    
  - **Gnocchi di nonna Alice:** Sold 1,263 units.
    
  - **Pavlova:** Quantity of 1,158 units sold.

- **Least Popular Products by Quantity Sold:**
  - **Mishi Kobe Niku:** Lowest quantity at 95 units.
    
  - **Genen Shouyu:** Sold 122 units.
    
  - **Gravad lax:** Total quantity of 125 units sold.
    
  - **Chocolade:** Quantity of 138 units sold.
    
  - **Laughing Lumberjack Lager:** Sold 184 units.

**Insight:**
Products with high total sales are key revenue drivers, while those with high quantities sold are popular among customers. Identifying and analyzing these products helps in optimizing inventory and marketing strategies.


### Which categories have the highest sales?

In [96]:
query = """
SELECT
    c.category_name,
    SUM(od.unit_price * od.quantity * (1 - od.discount)) AS TotalSales
FROM
    order_details od
JOIN
    products p ON od.product_id = p.product_id
JOIN
    categories c ON p.category_id = c.category_id
GROUP BY
    c.category_name
ORDER BY
    TotalSales DESC
"""
df = pd.read_sql_query(query,conn)
df


category_name     totalsales
0       Beverages  267868.179786
1  Dairy Products  234507.284531
2     Confections  167357.225473
3    Meat/Poultry  163022.360269
4         Seafood  131261.736556
5      Condiments  106047.084610
6         Produce   99984.580074
7  Grains/Cereals   95744.587355

In [97]:
# plotting
fig = px.bar(df,
            x='category_name',
            y='totalsales',
            title='Categories vs Sales',
            labels={'category_name':"Category Name",'totalsales':"Total Sales"},
            template='plotly_dark')
fig.show()

### Which Category Has the Highest Sales?

**Objective:** Determine the category with the highest total sales to understand which product categories are performing best.

**Findings:**

- **Beverages:** The highest total sales at approximately $267,868. This indicates that beverages are the top-performing category in terms of sales revenue.
    
- **Dairy Products:** Second highest with total sales of around $234,507, showing strong performance in dairy products.
    
- **Confections:** Generated approximately $167,357 in sales, making it the third highest category.
    
- **Meat/Poultry:** Total sales of about $163,022, placing it fourth in terms of sales revenue.
    
- **Seafood:** Sales amounting to $131,262, positioning it fifth among the top categories.
    
- **Condiments:** Total sales of around $106,047, ranking it sixth.
    
- **Produce:** Generated approximately $99,985 in sales, making it seventh.
    
- **Grains/Cereals:** The lowest among the categories with total sales of $95,745.

**Visualization:**
The bar graph illustrates the total sales for each category, clearly showing the dominant categories in terms of revenue.

**Insight:**
The Beverages category leads significantly in sales, which could be an area of focus for further investment and promotion. Understanding category performance helps in strategic planning and resource allocation to maximize profitability.


### How do product sales vary by supplier?

In [77]:
%%sql
SELECT
    s.company_name AS SupplierName,
    SUM(od.unit_price * od.quantity * (1 - od.discount)) AS TotalSales
FROM
    order_details od
JOIN
    products p ON od.product_id = p.product_id
JOIN
    suppliers s ON p.supplier_id = s.supplier_id
GROUP BY
    s.company_name
ORDER BY
    TotalSales DESC;


 * postgresql://postgres:***@localhost:5432/northwind
29 rows affected.


[('Aux joyeux ecclésiastiques', 153691.2755027201),
 ('Plutzer Lebensmittelgroßmärkte AG', 145372.39966947702),
 ('Gai pâturage', 117981.18020485654),
 ('Pavlova, Ltd.', 106459.77607838642),
 ("G'day, Mate", 65626.77016849845),
 ("Forêts d'érables", 61587.5698526758),
 ('Specialty Biscuits, Ltd.', 59032.0801634851),
 ('Pasta Buttini s.r.l.', 50254.609883744415),
 ('Formaggi Fortini s.r.l.', 48225.164374038046),
 ('Norske Meierier', 43141.5101294113),
 ('Leka Trading', 42017.64502122432),
 ("Grandma Kelly's Homestead", 41953.299975756556),
 ('Heli Süßwaren GmbH & Co. KG', 38653.419428995025),
 ('New Orleans Cajun Delights', 31167.989587372394),
 ('Tokyo Traders', 30526.33982929647),
 ('Karkki Oy', 28442.727384575122),
 ('New England Seafood Cannery', 26590.97431712504),
 ("Cooperativa de Quesos 'Las Cabras'", 25159.429822271763),
 ('Bigfoot Breweries', 22391.199810495975),
 ('Ma Maison', 22154.63772067912),
 ('Svensk Sjöföda AB', 20144.05991634145),
 ('Exotic Liquids', 19399.959904850424),
 ("Mayumi's", 14736.75505026741),
 ('Nord-Ost-Fisch Handelsgesellschaft mbH', 13424.197395977713),
 ('PB Knäckebröd AB', 11724.059940543175),
 ('Lyngbysild', 10221.175045646653),
 ('Escargots Nouveaux', 5881.6750394031405),
 ('Zaanse Snoepfabriek', 5326.792467173636),
 ('Refrescos Americanas LTDA', 4504.364968075454)]

### How Do Product Sales Vary by Supplier?

**Objective:** Analyze the total sales by supplier to understand which suppliers contribute most significantly to overall sales.

**Findings:**

- **Aux joyeux ecclésiastiques:** Highest total sales at approximately $153,691, making it the top-performing supplier.
  
- **Plutzer Lebensmittelgroßmärkte AG:** Second highest with total sales of around $145,372, indicating strong sales performance.
  
- **Gai pâturage:** Generated approximately $117,981 in sales, ranking third among suppliers.
  
- **Pavlova, Ltd.:** Total sales of about $106,460, placing it fourth in terms of sales revenue.
  
- **G'day, Mate:** Sales amounting to $65,627, positioning it fifth.
  
- **Forêts d'érables:** Generated approximately $61,588 in sales, ranking sixth.
  
- **Specialty Biscuits, Ltd.:** Total sales of around $59,032, making it seventh.
  
- **Pasta Buttini s.r.l.:** Sales of about $50,255, placing it eighth.
  
- **Formaggi Fortini s.r.l.:** Total sales of around $48,225, ranking ninth.
  
- **Norske Meierier:** Sales amounting to $43,142, making it tenth among the top suppliers.

**Insight:**

The supplier "Aux joyeux ecclésiastiques" leads in total sales, followed closely by "Plutzer Lebensmittelgroßmärkte AG." Understanding supplier performance can help in managing supplier relationships and focusing on high-performing suppliers for future growth strategies.


## Conclusion

Our analysis provides valuable insights into sales performance and customer behavior:

- **Total Sales by Year:** Sales grew significantly from 1996 to 1997, with a slight decrease in 1998.
  
- **Top Revenue-Generating Products:** Côte de Blaye and Thüringer Rostbratwurst are top revenue generators, while Chocolade and Geitost have the lowest sales.
  
- **Top Customers by Sales Amount:** QUICK-Stop and Save-a-lot Markets are our highest revenue-generating customers.
  
- **Sales Trends by Region:** The USA and Germany lead in sales, followed by Austria, Brazil, and France.
  
- **Customer Order Frequency:** Save-a-lot Markets and Ernst Handel place the most orders.
  
- **Average Order Value:** QUICK-Stop and SIMOB have the highest average order values after discount.
  
- **Product Popularity:** Camembert Pierrot and Raclette Courdavault are the most popular products, while Mishi Kobe Niku and Genen Shouyu are the least.
  
- **Category Sales Performance:** Beverages and Dairy Products have the highest total sales.
  
- **Supplier Contribution:** Aux joyeux ecclésiastiques and Plutzer Lebensmittelgroßmärkte AG are the top suppliers by sales.

### Recommendations

1. **Focus on High-Performing Products and Categories:** Boost marketing and inventory for top products and categories.
2. **Enhance Customer Engagement:** Develop loyalty programs for top customers.
3. **Expand in High-Growth Regions:** Allocate resources for the USA and Germany.
4. **Strengthen Supplier Relationships:** Collaborate with high-performing suppliers for better deals and availability.
